### PR TITLE
OCPBUGS-38584: Authentication metrics wasn't collected with 4.15+

### PR DESCRIPTION
**Pull request by Lucifergene**
Jira: https://issues.redhat.com/browse/OCPBUGS-38584

Solution Description:

Although the metrics are present, they were not getting called. 
In this PR, we are making sure they get called to be able to fetch those metrics and their correct values.

image

### CODE REVIEW

File: pkg/auth/metrics.go [hunk 1]
```diff
@@ -79,19 +80,26 @@ func (m *Metrics) loginSuccessfulSync(k8sConfig *rest.Config, ls *sessions.Login
...
 		Timeout:     30 * time.Second,
 	}
 
 	role := UnknownLoginRole
 
-	if isKubeAdmin, err := m.isKubeAdmin(ctx, configWithBearerToken); isKubeAdmin && err == nil {
+	if isKubeAdmin, err := m.isKubeAdmin(ctx, anonConfigWithBearerToken); isKubeAdmin && err == nil {
 		role = KubeadminLoginRole
-	} else if canGetNamespaces, err := m.canGetNamespaces(ctx, configWithBearerToken); err == nil {
+	} else if canGetNamespaces, err := m.canGetNamespaces(ctx, anonConfigWithBearerToken); err == nil {
```
Comment (jerolimov) [inline pkg/auth/metrics.go:102]: Is this an anonymous bearer token? :thinking:

File: pkg/auth/metrics.go [hunk 2]
```diff
@@ -43,11 +43,12 @@ const (
...
-	loginSuccessful      *prometheus.CounterVec
-	loginFailures        *prometheus.CounterVec
-	logoutRequests       *prometheus.CounterVec
-	tokenRefreshRequests *prometheus.CounterVec
+	loginRequests             prometheus.Counter
+	loginSuccessful           *prometheus.CounterVec
+	loginFailures             *prometheus.CounterVec
+	logoutRequests            *prometheus.CounterVec
+	tokenRefreshRequests      *prometheus.CounterVec
+	internalproxyClientConfig *rest.Config
```
Comment (jerolimov) [inline pkg/auth/metrics.go:51]: Reminder for myself: I need to understand why this is now required and not before.

File: pkg/auth/oauth2/auth_oidc_test.go [hunk 3]
```diff
@@ -39,6 +40,10 @@ const (
 	testNewRefreshToken   = "new-refresh-token"
 )
 
+var (
+	internalproxyClientConfig = &rest.Config{}
```
Comment (jerolimov) [inline pkg/auth/oauth2/auth_oidc_test.go:44]: uppercase proxy?

```suggestion
	internalProxyClientConfig = &rest.Config{}
```

File: pkg/auth/metrics_test.go [hunk 4]
```diff
@@ -13,8 +13,12 @@ import (
 	"k8s.io/client-go/rest"
 )
 
+var (
+	internalproxyClientConfig = &rest.Config{}
```
Comment (jerolimov) [inline pkg/auth/metrics_test.go:17]: uppercase proxy? In this case you can also define that it doesn't configure anything like:

```suggestion
	defaultRestClientConfig = &rest.Config{}
```

File: pkg/auth/metrics.go [hunk 5]
```diff
@@ -79,10 +80,17 @@ func (m *Metrics) loginSuccessfulSync(k8sConfig *rest.Config, ls *sessions.Login
...
+	if err != nil {
+		klog.Errorf("failed to set up an anonymous roundtripper: %v", err)
+		return
+	}
+
 	ctx := context.TODO()
+
 	configWithBearerToken := &rest.Config{
 		Host:        k8sConfig.Host,
-		Transport:   k8sConfig.Transport,
```
Comment (Lucifergene) [inline pkg/auth/metrics.go:85]: `k8sConfig.Transport` is nil.

This the entire k8sConfig:

```
I0828 11:02:11.879264 1 metrics.go:114] auth.Metrics loginSuccessfulSync - k8sConfig: &rest.Config{Host:"https://kubernetes.default.svc", APIPath:"", ContentConfig:rest.ContentConfig{AcceptContentTypes:"", ContentType:"", GroupVersion:(*schema.GroupVersion)(nil), NegotiatedSerializer:runtime.NegotiatedSerializer(nil)}, Username:"", Password:"", BearerToken:"", BearerTokenFile:"/var/run/secrets/kubernetes.io/serviceaccount/token", Impersonate:rest.ImpersonationConfig{UserName:"", UID:"", Groups:[]string(nil), Extra:map[string][]string(nil)}, AuthProvider:<nil>, AuthConfigPersister:rest.AuthProviderConfigPersister(nil), ExecProvider:<nil>, TLSClientConfig:rest.sanitizedTLSClientConfig{Insecure:false, ServerName:"", CertFile:"", KeyFile:"", CAFile:"/var/run/secrets/kubernetes.io/serviceaccount/ca.crt", CertData:[]uint8(nil), KeyData:[]uint8(nil), CAData:[]uint8(nil), NextProtos:[]string(nil)}, UserAgent:"", DisableCompression:false, Transport:http.RoundTripper(nil), WrapTransport:(transport.WrapperFunc)...
```

From here, you can see that `Transport:http.RoundTripper(nil)`

Therefore we needed to create the Transport.

Comment (jhadvig) [reply pkg/auth/metrics.go:85]: lets pass the already instantiated `anonymousInternalProxiedK8SRT` from the server.go

**Review by jhadvig — APPROVED** [review]
/approve
/lgtm

**Pull request merged**